### DISCUSSION TITLE
When using Quasar + Pinia + Boot file it gives a console browser warning

### DISCUSSION THREAD

**Discussion by BenJackGill** [quasarframework/quasar#12367] [General - Components / Directives / etc] [answered]
I am using Quasar + Pinia + TypeScript.

And I want to use a boot file to load the user information into the user store when the app loads.

Everything works as expected, but the browser console is throwing a warning. Maybe it has to do with my setup.

1) Installed Pinia using `npm install pinia@next`.
2) Created `src/store/index.ts` with the code below:
```
import { store } from 'quasar/wrappers';
export default store;
```
3) Added a `pinia.ts` boot file as below:
```
import { boot } from 'quasar/wrappers';
import { createPinia } from 'pinia';
const pinia = createPinia();
export default boot(({ app }) => {
  app.use(pinia);
});
export { pinia };
```
4) Added `pinia` as a boot file in `quasar.conf.js` as below:
```
boot: [
      'pinia',
    ],
```
5) Created an empty user store in `src/store/user.ts` as per below:
```
import { User } from 'firebase/auth';
import { defineStore } from 'pinia';
export const useUserStore = defineStore('user', {
  state: () =>
    ({
      // User properties and values go here. I removed them for this example.
    } as unknown as User),
});
```
6) Created a new boot file `user.ts` to add the user information to the user store on app load:
```
import { User } from 'firebase/auth';
import { boot } from 'quasar/wrappers';
import { checkAuthStatus } from 'src/firebase/config';
import { useUserStore } from 'src/store/user';
export default boot(async (/*{ store }*/) => {
  // Check for user
  const user = (await checkAuthStatus()) as User;
  // Add user to store
  if (user) {
    const userStore = useUserStore();
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    userStore.$state = user;
    console.log(userStore);
  }
});
```
7) Added `user` as a boot file in `quasar.conf.js` as below:
```
boot: [
      'pinia',
      'user',
    ],
```

Everything seems to work fine. The user info is stored correctly.

But when adding the `user.ts` boot file (steps 6 and 7) the browser console shows this warning message:
`[Vue warn]: A plugin must either be a function or an object with an "install" function.`

Any idea what's wrong?

I like to clear all warning messages instead of letting them build up, even if it has no effect on the functioning of the app.

**Comment by yusufkandemir** [comment, marked as answer]
You can either (1) get rid of the Pinia poot file and return the Pinia instance from the function you define in `store/index.[js|ts]` or (2) completely remove `store/index.[js|ts]` and keep your current structure.

For (1), you can check out https://github.com/TobyMosque/quasar-v2-ssr-pinia/blob/master/src/store/index.ts

**Reply by BenJackGill** [reply]
Thanks for the help.

**Option 1:**

I'm not using SSR so I tried to "directly export the Store instantiation" as advised in that linked file. Not sure if I've done that correctly, because I get this new warning:

```
[Quasar] boot error: Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?

const pinia = createPinia()
app.use(pinia)

This will fail in production.
    at useStore (pinia.esm-bundler.js?ec33:1561:1)
    at eval (user.ts?857a:9:1)
    at Generator.next (<anonymous>)
    at fulfilled (user.ts:11:58)
```

Here is my attempt at directly exporting the Store:

```
import { store } from 'quasar/wrappers';
import { createPinia, Pinia } from 'pinia';
import { unref, Ref } from 'vue';
// import example from './module-example'
// import { ExampleStateInterface } from './module-example/state';

/*
 * If not building with SSR mode, you can
 * directly export the Store instantiation;
 *
 * The function below can be async too; either use
 * async/await or return a Promise which resolves
 * with the Store instance.
 */

declare module '@quasar/app' {
  interface QSsrContext {
    state: Ref<never> | never;
  }
}

// provide typings for `this.$store`
declare module '@vue/runtime-core' {
  interface ComponentCustomProperties {
    $store: Pinia;
  }
}

declare module 'pinia' {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  interface Pinia {
    replaceState(state: never): void;
  }
}

export default store;
```

I also tried changing that last line and exporting like this, but I get the same error Quasar/getActivePinia error as above:

```
export default store(function () {
  const pinia = createPinia();
  return pinia;
});
```

**Option 2:**

When keeping  the current structure and removing the `store/index.[js|ts]` file I get this error:

```
App •  ERROR  •  UI  in ./.quasar/app.js
Module not found: Can't resolve imported dependency "app/src/store/index"
Did you forget to install it? You can run: npm install --save app/src/store/index
```

**Reply by yusufkandemir** [reply]
*Option 1:*
This can't work:
```js
import { store } from 'quasar/wrappers';
export default store;
```
because you are not doing anything related to Pinia. You are just exporting an empty function, as `store` is a "wrapper" function.

The store is initialized before boot files, so I am pretty sure you have some other stuff somewhere because you seem to get a bit overwhelmed. You don't need to worry about removing SSR stuff, your code will still work. I would recommend mimicking the structure and getting it to work first. Then, you can make modifications if you wish.

**Option 2:**
Sorry, a correction: you must get rid of the `store` folder completely, not just the `index.[js|ts]` file.

**Reply by BenJackGill** [reply]
I only included those two lines in the `index.ts` file to stop Quasar errors from showing. It doesn't like it when there is no `src/store/index.ts` file and that was the simplest `index.ts` I could think of.

Also if I keep the SSR content of the `index.ts` you linked above it throws a bunch of TypeScript and EsLint errors. So I tried exporting it directly as advised in the file.

**Reply by yusufkandemir** [reply]
I guess we will not make a lot of progress like this. Please share a repro, so I can reproduce the errors, see your code, and try some solutions.

**Reply by BenJackGill** [reply]
Thanks! I really appreciate your help with this.

I have made the repo public temporarily.

https://github.com/BenJackGill/agentcake-app

**Reply by yusufkandemir** [reply]
I found the explanations and examples a bit confusing and contradicting here: https://pinia.vuejs.org/core-concepts/outside-component-usage.html
But, it definitely works when you provide the Pinia instance explicitly. So, to sum up everything, I got it working without errors or warnings by making the changes below:

```diff
diff --git a/src/store/index.ts b/src/store/index.ts
index a9620d2..570355b 100644
--- a/src/store/index.ts
+++ b/src/store/index.ts
@@ -1,6 +1,5 @@
 import { store } from 'quasar/wrappers';
 import { createPinia, Pinia } from 'pinia';
-import { unref, Ref } from 'vue';
 // import example from './module-example'
 // import { ExampleStateInterface } from './module-example/state';
 
@@ -14,8 +13,9 @@ import { unref, Ref } from 'vue';
  */
 
 declare module '@quasar/app' {
-  interface QSsrContext {
-    state: Ref<never> | never;
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface BootFileParams<TState> {
+    store: Pinia;
   }
 }
 
@@ -26,26 +26,8 @@ declare module '@vue/runtime-core' {
   }
 }
 
-declare module 'pinia' {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
-  interface Pinia {
-    replaceState(state: never): void;
-  }
-}
-
-export default store(function ({ ssrContext }) {
+export default store(function () {
   const pinia = createPinia();
-  if (process.env.SERVER && ssrContext) {
-    ssrContext.onRendered(function () {
-      // unwrapping the state for serialization
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-      ssrContext.state = unref(ssrContext.state);
-    });
-  }
-  if (process.env.CLIENT) {
-    pinia.replaceState = function (state: never) {
-      pinia.state.value = state;
-    };
-  }
+
   return pinia;
 });
```

```diff
diff --git a/src/boot/user.ts b/src/boot/user.ts
index 9a66c63..80c547f 100644
--- a/src/boot/user.ts
+++ b/src/boot/user.ts
@@ -3,10 +3,10 @@ import { boot } from 'quasar/wrappers';
 import { checkAuthStatus } from 'src/firebase/config';
 import { useUserStore } from 'src/store/user';
 
-export default boot(async (/*{ store }*/) => {
+export default boot(async ({ store }) => {
   const user = (await checkAuthStatus()) as User;
   if (user) {
-    const userStore = useUserStore();
+    const userStore = useUserStore(store);
     userStore.setUser(user);
   }
 });
```

There were some minimal type errors in your code:
```diff
diff --git a/src/pages/BillingPage.vue b/src/pages/BillingPage.vue
index 4b98b9e..8147978 100644
--- a/src/pages/BillingPage.vue
+++ b/src/pages/BillingPage.vue
@@ -8,7 +8,8 @@
 
 <script setup lang="ts">
 import { useUserStore } from 'src/store/user';
-const user = useUserStore();
+const userStore = useUserStore();
+const user = userStore.user;
 const name = user.displayName;
 const email = user.email;
 </script>
```

**Reply by BenJackGill** [reply]
You. Are. AMAZING.

Thank you so much. I have been struggling with this for a while now.
